File: ontrack-ui/src/test/java/net/nemerosa/ontrack/boot/ui/ProjectControllerIT.java
package net.nemerosa.ontrack.boot.ui;

import net.nemerosa.ontrack.model.security.ProjectCreation;
import net.nemerosa.ontrack.model.security.ProjectEdit;
import net.nemerosa.ontrack.model.structure.ID;
import net.nemerosa.ontrack.model.structure.NameDescriptionState;
import net.nemerosa.ontrack.model.structure.Project;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.AccessDeniedException;

import static org.junit.Assert.*;

public class ProjectControllerIT extends AbstractWebTestSupport {

    @Autowired
    private ProjectController controller;

    @Test
    public void createProject() throws Exception {
        asUser().with(ProjectCreation.class).call(() -> {
            NameDescriptionState nameDescription = nameDescription().asState();
            Project resource = controller.newProject(nameDescription);
            checkProject(resource, nameDescription);
            return null;
        });
    }

    @Test(expected = AccessDeniedException.class)
    public void createProject_denied() throws Exception {
        asUser().call(() -> controller.newProject(nameDescription().asState()));
    }

    @Test
    public void disablingEnablingProject() throws Exception {
        Project project = doCreateProject();
        // Disables it
        Project disabled = asUser().with(project, ProjectEdit.class).call(
                () -> controller.disableProject(project.getId())
        );
        assertTrue("Project is disabled", disabled.isDisabled());
        // Enables it
        Project enabled = asUser().with(project, ProjectEdit.class).call(
                () -> controller.enableProject(project.getId())
        );
        assertFalse("Project is enabled", enabled.isDisabled());
    }

    @Test
    public void updateProject() throws Exception {
        // Creates the project
        NameDescriptionState initialNames = nameDescription().asState();
        Project project = asUser().with(ProjectCreation.class).call(() -> controller.newProject(initialNames));
        ID id = project.getId();
        // Edition
        asUser().with(id.getValue(), ProjectEdit.class).call(() -> {
            // Updates
            NameDescriptionState nameDescription = nameDescription().asState();
            assertNotEquals(initialNames, nameDescription);
            Project updated = controller.saveProject(id, nameDescription);
            // Checks
            checkProject(updated, nameDescription);
            // Gets the project back
            updated = controller.getProject(id);
            checkProject(updated, nameDescription);
            return null;
        });
    }

    private void checkProject(Project project, NameDescriptionState nameDescription) {
        assertNotNull("Project not null", project);
        assertNotNull("Project ID not null", project.getId());
        assertTrue("Project ID set", project.getId().isSet());
        assertEquals("Project name", nameDescription.getName(), project.getName());
        assertEquals("Project description", nameDescription.getDescription(), project.getDescription());
        assertEquals("Project state", nameDescription.isDisabled(), project.isDisabled());
    }

}

File: ontrack-ui/src/test/java/net/nemerosa/ontrack/boot/ui/ProjectControllerIT.kt
package net.nemerosa.ontrack.boot.ui

import net.nemerosa.ontrack.model.events.EventFactory
import net.nemerosa.ontrack.model.events.EventQueryService
import net.nemerosa.ontrack.model.security.ProjectCreation
import net.nemerosa.ontrack.model.security.ProjectEdit
import net.nemerosa.ontrack.model.structure.NameDescriptionState
import net.nemerosa.ontrack.model.structure.Project
import net.nemerosa.ontrack.model.structure.ProjectEntityType
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.access.AccessDeniedException
import kotlin.test.*

class ProjectControllerIT : AbstractWebTestSupport() {

    @Autowired
    private lateinit var controller: ProjectController

    @Autowired
    private lateinit var eventQueryService: EventQueryService

    @Test
    fun createProject() {
        asUser().with(ProjectCreation::class.java).execute {
            val nameDescription = nameDescription().asState()
            val resource = controller.newProject(nameDescription)
            checkProject(resource.body!!, nameDescription)
        }
    }

    @Test
    fun createProject_denied() {
        assertFailsWith<AccessDeniedException> {
            asUser().call { controller.newProject(nameDescription().asState()) }
        }
    }

    @Test
    fun disablingEnablingProject() {
        val project = doCreateProject()
        // Disables it
        val disabled = asUser().withProjectFunction(project, ProjectEdit::class.java).call { controller.disableProject(project.id) }.body!!
        assertTrue(disabled.isDisabled, "Project is disabled")
        val disabledEvent = asUser { eventQueryService.getLastEvent(project, EventFactory.DISABLE_PROJECT) }
        assertNotNull(disabledEvent, "Disabled event is there") { event ->
            val eventProject: Project = event.getEntity(ProjectEntityType.PROJECT)
            assertNotNull(eventProject, "Project associated to the event") {
                assertEquals(project.name, it.name)
            }
        }
        // Enables it
        val enabled = asUser().withProjectFunction(project, ProjectEdit::class.java).call { controller.enableProject(project.id) }.body!!
        assertFalse(enabled.isDisabled, "Project is enabled")
        val enabledEvent = asUser { eventQueryService.getLastEvent(project, EventFactory.ENABLE_PROJECT) }
        assertNotNull(enabledEvent, "Enabled event is there") { event ->
            val eventProject: Project = event.getEntity(ProjectEntityType.PROJECT)
            assertNotNull(eventProject, "Project associated to the event") {
                assertEquals(project.name, it.name)
            }
        }
    }

    @Test
    fun updateProject() {
        // Creates the project
        val initialNames = nameDescription().asState()
        val project = asUser().with(ProjectCreation::class.java).call { controller.newProject(initialNames) }.body!!
        val id = project.id
        // Edition
        asUser().withProjectFunction(project, ProjectEdit::class.java).execute {
            // Updates
            val nameDescription = nameDescription().asState()
            assertNotEquals(initialNames, nameDescription)
            var updated = controller.saveProject(id, nameDescription).body!!
            // Checks
            checkProject(updated, nameDescription)
            // Gets the project back
            updated = controller.getProject(id).body!!
            checkProject(updated, nameDescription)
        }
    }

    private fun checkProject(project: Project, nameDescription: NameDescriptionState) {
        assertNotNull(project, "Project not null")
        assertNotNull(project.id, "Project ID not null")
        assertTrue(project.id.isSet, "Project ID set")
        assertEquals(nameDescription.name, project.name, "Project name")
        assertEquals(nameDescription.description, project.description, "Project description")
        assertEquals(nameDescription.isDisabled, project.isDisabled, "Project state")
    }

}
